ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -136,15 +136,18 @@ private void processHead(Optional<UInt64> nodeSlot) {
   public SafeFuture<AttestationProcessingResult> onAttestation(
       final ValidateableAttestation attestation) {
     return recentChainData
-        .retrieveCheckpointState(attestation.getData().getTarget())
+        .retrieveBlockState(attestation.getData().getBeacon_block_root())
```

Comment:
This is going to destroy us performance-wise.   We will suddenly be missing checkpoint state caches a whole lot more.

Given that the target checkpoint must be for the epoch matching the slot we should continue using the state for the target state.  ie from `ForkChoiceUtil.validateOnAttestation`:
```
    if (!target.getEpoch().equals(compute_epoch_at_slot(attestation.getData().getSlot()))) {
      return AttestationProcessingResult.invalid("Attestation slot must be within specified epoch");
    }
```

Remember that if the randao in the state we validate against was different to the one used when creating the attestation, the signature would be invalid because the committees wouldn't match up.  So we can safely validate against any state from one epoch prior to the attestation's slot.